### ISSUE TITLE
A better visibility control?

### ISSUE THREAD

**Issue by Evian-Zhang** [m4b/goblin#233]
As Rust supports a large variety of [visibility and privacy](https://doc.rust-lang.org/reference/visibility-and-privacy.html) modifiers, I think it would be better if we use more specific visibility modifier beyond `pub` and not `pub`.

For example, the `peek_bytes` and `peek` method in `src/lib.rs` may could be marked as `pub(crate)` or even private, and the `is_in_range` and other methods in `src/pe/utils.rs` may could be marked as `pub(crate)`.

By doing this, we could get a more clear rustdoc for users to use this library.

**Comment by sollyucko** [comment]
FWIW, `#[doc(hidden)]` can be used to remove the methods from the rustdoc while leaving them publicly accessible. (See https://doc.rust-lang.org/rustdoc/the-doc-attribute.html#dochidden, https://github.com/rust-lang/api-guidelines/issues/27, and https://rust-lang.github.io/api-guidelines/documentation.html#rustdoc-does-not-show-unhelpful-implementation-details-c-hidden)

**Comment by Evian-Zhang** [comment]
> FWIW, `#[doc(hidden)]` can be used to remove the methods from the rustdoc while leaving them publicly accessible. (See https://doc.rust-lang.org/rustdoc/the-doc-attribute.html#dochidden, [rust-lang/api-guidelines#27](https://github.com/rust-lang/api-guidelines/issues/27), and https://rust-lang.github.io/api-guidelines/documentation.html#rustdoc-does-not-show-unhelpful-implementation-details-c-hidden)

In my understanding, the `#[doc(hidden)]` macro is used for hide the unhelpful `impl` message, which could not be hidden by visibility modifier.

And for those methods which need to be publicly accessible, we can just use `pub` modifier. What I mean is to change to a more specific visibility modifier for those are only needed within this crate, and is useless for people using this crate as library.

**Comment by m4b** [comment]
Something like this sounds good to me. Would you @Evian-Zhang (or anyone else) be interested in collecting a list of pub functions that are candidates for pub(crate)?

Please note: all pub to pub(crate) changes are breaking so be aware that they could be used, and that we’ll need to think carefully whether to go forward.

Although as you said, I think some of the functions / types are basically crate private and not used by anyone (most of the original types and functions predates pub(crate) addition to the language !)

**Comment by Evian-Zhang** [comment]
> Something like this sounds good to me. Would you @Evian-Zhang (or anyone else) be interested in collecting a list of pub functions that are candidates for pub(crate)?
> 
> Please note: all pub to pub(crate) changes are breaking so be aware that they could be used, and that we’ll need to think carefully whether to go forward.
> 
> Although as you said, I think some of the functions / types are basically crate private and not used by anyone (most of the original types and functions predates pub(crate) addition to the language !)

I agree that this change can be very dangerous because it is breaking. So it needs the writers of this library and writers of some famous libraries using goblin together to change the previous-written visibility modifier.

So I suggest taking a long-term plan to change those already-written code, and for new PRs, take a close look at the visibility modifiers before merging it.
